src/core/memory.d

```diff
@@ -919,3 +919,218 @@ extern (C) private pure @system @nogc nothrow
 
     pragma(mangle, "free") void fakePureFree(void* ptr);
 }
+
+extern(C) private @system nothrow
+{
+    pragma(mangle, "_d_delinterface") void _d_delinterface(void**);
+    pragma(mangle, "_d_delclass") void _d_delclass(Object*);
+    pragma(mangle, "_d_delstruct") void _d_delstruct(void**, TypeInfo_Struct);
+    pragma(mangle, "_d_delmemory") void _d_delmemory(void**);
+    pragma(mangle, "_d_delarray_t") void _d_delarray_t(void**, TypeInfo_Struct);
+}
+
+/**
+The `delete` keyword allowed to free GC-allocated memory.
+As this is inherently not `@safe`, it has been deprecated.
+This function has been added to provide an easy transition from `delete`.
+It performs the same functionality as `delete`.
+
+Note: Users should prefer $(REF destroy, object)` to explicitly finalize objects,
+and only resort to $(REF __delete, core,memory) when $(REF destroy, object)
+would not be a feasible option.
+
+In detail, `__delete(x)` returns with no effect if `x` is `null`. Otherwise, it
+performs the following actions in sequence:
+$(UL
+    $(LI
+        Calls `.object.destroy(x)` (if `x` is a class or interface object) or
+        `typeid(*x).destroy(x)` (if `x` is pointer to a struct) to destroy the referred entity.
```

Comment:
Here it should just be `.object.destroy(*x)`, no? It's faster and better documented.

Reply:
Are you sure about `object.destroy` being faster?

`TypeInfo_Struct.destroy`:

https://github.com/dlang/druntime/blob/18be8bf97933c2e5c0d0948f18477268ac7c1766/src/object.d#L1192-L1201

`object.destroy`:

https://github.com/dlang/druntime/blob/18be8bf97933c2e5c0d0948f18477268ac7c1766/src/object.d#L2954-L2965

Reply:
The former is an indirect call and the second is a direct call. Truth be told, destroy does more work because it reinitializes the object. In the case of `__delete` we don't care to reinitialize the object because we just wrote in the documentation that access after `__delete` is undefined behavior.

So all in all the best way to go is to write in the documentation that we call `destroy` and actually call `_destructRecurse`. That covers all defined cases and is fast.

Reply:
`_destructRecurse` is private and in `object` - so we can't even do `package(core)`.